oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSuggestIterator.java

```diff
@@ -85,11 +87,17 @@ private void loadSuggestions() throws IOException {
         SearchRequest searchRequest = new SearchRequest(indexNode.getDefinition().getRemoteIndexAlias())
                 .source(searchSourceBuilder);
         SearchResponse res = indexNode.getConnection().getClient().search(searchRequest, RequestOptions.DEFAULT);
+        List<String> suggestionList = new ArrayList<>();
         PriorityQueue<ElasticSuggestion> suggestionPriorityQueue = new PriorityQueue<>((a, b) -> Double.compare(b.score, a.score));
         for (SearchHit doc : res.getHits()) {
             if (responseHandler.isAccessible(responseHandler.getPath(doc))) {
                 for (SearchHit suggestion : doc.getInnerHits().get(FieldNames.SUGGEST).getHits()) {
-                    suggestionPriorityQueue.add(new ElasticSuggestion((String) suggestion.getSourceAsMap().get("value"), suggestion.getScore()));
+                    String suggestValue = (String) suggestion.getSourceAsMap().get("value");
+                    // Avoid adding duplicate suggest results
+                    if (!suggestionList.contains(suggestValue)) {
+                        suggestionPriorityQueue.add(new ElasticSuggestion((String) suggestion.getSourceAsMap().get("value"), suggestion.getScore()));
+                        suggestionList.add(suggestValue);
+                    }
```

Comment:
this can be simplified. I would suggest rolling back the changes in this file and:

- change the last instruction with `this.internalIterator = suggestionPriorityQueue.stream().distinct().iterator();`
- override `equals` and `hashCode` in `ElasticSuggestion` to look only at `suggestion` field

This would:

- remove the additional `List` to keep track of the suggestions
- simplify the code
- avoid cases where the same suggestion with a lower score comes before one with a higher one (not sure if this is actually possible)

Reply:
Yes, there are two problems with using an ArrayList:

* suggestionList.contains(suggestValue) can become a performance problem if there are many entries: with 1000 entries, 1 million comparisons are needed: O(n^2), same as bubble sort.
* In your implementation, the first entry is added, and the second entry it ignored, even if it has a higher score... not sure if this can happen in practise.

Using suggestionPriorityQueue.stream().distinct().iterator() should resolve both issues.

Reply:
@fabriziofortino  @thomasmueller - made the changes.